### ISSUE TITLE
[v1.6.0] Release Tracker

### ISSUE THREAD

**Issue by seemethere** [pytorch/pytorch#40472]
We cut a [release branch](https://github.com/pytorch/pytorch/tree/release/1.6) for the 1.6.0 release

This issue is for tracking cherry-picks to the branch.  The criteria for a change to be included is ALL of the following:

1. bug fix or documentation fix/improvement (i.e. NOT a feature and doesn't break BC)
2. already landed in master or specific to 1.6.0
3. worth the risk/reward.  We will most likely only include changes that fix regressions from 1.5.1 to 1.6.0 (or fixes for features that weren't in 1.5), but tiny fixes for large bugs will also be considered.

If you need a change to be cherry-picked to the branch, please open a PR against the [release/1.6](https://github.com/pytorch/pytorch/tree/release/1.6) branch and comment below.  Any details you can provide about the criteria above (e.g. links to master PR, description of the severity of the bug fix) are helpful.  Ultimately we will make an inclusion / no inclusion call and communicate that here.

**NOTE: do not land the PRs yourself.  Our normal tools (ghstack / ghimport, etc.) do not work on the release branch.  Someone else will land your change for you.**

--------

[Current PRs open against the branch](https://github.com/pytorch/pytorch/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+base%3Arelease/1.6)

cc @ezyang @seemethere @malfet @gchanan

**Comment by jerryzh168** [comment]
qq: looks like the branch cut is done before 5pm? my PR is landed before 5pm but is not included.

**Comment by seemethere** [comment]
> qq: looks like the branch cut is done before 5pm? my PR is landed before 5pm but is not included.

Branch cut was done off of `viable/strict` so your PR must've landed during a time where the branch was not stable.

**Comment by jerryzh168** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40476

**Comment by jlin27** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40482

**Comment by z-a-f** [comment]
Please merge #40481

**Comment by peterjc123** [comment]
Please merge https://github.com/pytorch/pytorch/pull/40487

**Comment by hx89** [comment]
Please merge #40475

**Comment by mrshenli** [comment]
placeholder for #40461

**Comment by dreiss** [comment]
Requesting merge of #37243

**Comment by Laksh1997** [comment]
When will pytorch 1.6 be released?

**Comment by wconstab** [comment]
Please merge #40538

**Comment by mruberry** [comment]
Please merge https://github.com/pytorch/pytorch/pull/40542, which fixes #40140.

**Comment by jlin27** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40540

**Comment by jlin27** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40539

**Comment by eellison** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40569 needed for feature release

**Comment by eellison** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40572 to fix torch hub for torchvision

**Comment by jerryzh168** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40579

**Comment by mrshenli** [comment]
placeholder for #40614

cc @lw

**Comment by jerryzh168** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40647

**Comment by peterjc123** [comment]
Please merge: https://github.com/pytorch/pytorch/pull/40676

**Comment by mrshenli** [comment]
Please merge: #40688

**Comment by wanchaol** [comment]
Please merge https://github.com/pytorch/pytorch/pull/40582 https://github.com/pytorch/pytorch/pull/40424 and https://github.com/pytorch/pytorch/pull/39601 (https://github.com/pytorch/pytorch/pull/40424 should come before https://github.com/pytorch/pytorch/pull/39601)

**Comment by seemethere** [comment]
> Please merge #40582 #40424 and #39601 (#40424 should come before #39601)

Would you mind making the cherry pick PRs to `release/1.6`?

**Comment by wanchaol** [comment]
> > Please merge #40582 #40424 and #39601 (#40424 should come before #39601)
> 
> Would you mind making the cherry pick PRs to `release/1.6`?

Yes sorry deleted that comment lol. I will create those cherry pick PRs :)

**Comment by eellison** [comment]
please merge https://github.com/pytorch/pytorch/pull/40645 (shape analysis fix for legacy fuser)

**Comment by malfet** [comment]
> please merge #40645 (shape analysis fix for legacy fuser)
Please add corresponding PR/commit hash against master branch to the PR description.

**Comment by jerryzh168** [comment]
Please merge:
https://github.com/pytorch/pytorch/pull/40748
https://github.com/pytorch/pytorch/pull/40750
https://github.com/pytorch/pytorch/pull/40751
https://github.com/pytorch/pytorch/pull/40752

**Comment by jamesr66a** [comment]
Please merge https://github.com/pytorch/pytorch/pull/40757

**Comment by anjali411** [comment]
Please merge this doc fix: https://github.com/pytorch/pytorch/pull/40450

**Comment by jamesr66a** [comment]
Please merge https://github.com/pytorch/pytorch/pull/40793

**Comment by anjali411** [comment]
please merge: https://github.com/pytorch/pytorch/pull/40513

**Comment by anjali411** [comment]
please merge: https://github.com/pytorch/pytorch/pull/39087

**Comment by wanchaol** [comment]
Please merge https://github.com/pytorch/pytorch/pull/40796 and https://github.com/pytorch/pytorch/pull/40797 (failures seems not related)

**Comment by ZolotukhinM** [comment]
Please merge #40716

**Comment by mruberry** [comment]
Please merge #40799

**Comment by peterjc123** [comment]
Please merge https://github.com/pytorch/pytorch/pull/40849.

**Comment by peterjc123** [comment]
Please merge #40852.

**Comment by malfet** [comment]
> please merge: #40513

Please create a cherry-pick PR

**Comment by seemethere** [comment]
Letting everyone know here that we will be cutting an RC2, any cherry picks below this comment will not be included in the RC2

**Comment by AshkanAliabadi** [comment]
> Requesting merge of #37243
> 
> @dreiss This looks like a really big and risky change an it seems to have caused one regression already (see #40668 ) Does it really have to go into 1.6?

Please merge the following patches in the order below.  It is all or none situation, including the one David linked to:

0) https://github.com/pytorch/pytorch/pull/37243
1)  https://github.com/pytorch/pytorch/commit/bdc00196d1b63a339a5b9b6aaec44cefe069a6ea (no PR - originated from a Phabricator diff)
2) https://github.com/pytorch/pytorch/pull/40522
3) https://github.com/pytorch/pytorch/pull/40520
4) https://github.com/pytorch/pytorch/pull/40516
5) https://github.com/pytorch/pytorch/pull/40526
6) https://github.com/pytorch/pytorch/pull/40707

@dreiss

**Comment by dreiss** [comment]
Requesting #40587

**Comment by ezyang** [comment]
Please merge https://github.com/pytorch/pytorch/pull/40881

**Comment by seemethere** [comment]
> > Requesting merge of #37243
> > @dreiss This looks like a really big and risky change an it seems to have caused one regression already (see #40668 ) Does it really have to go into 1.6?
> 
> Please merge the following patches in the order below. It is an all or none situation, including the one David linked to:
> 
> 1. #37243
> 2. [bdc0019](https://github.com/pytorch/pytorch/commit/bdc00196d1b63a339a5b9b6aaec44cefe069a6ea) (no PR - originated from a Phabricator diff)
> 3. #40522
> 4. #40520
> 5. #40516
> 6. #40526
> 7. #40707
> 
> @dreiss

Could you elaborate on what this actually does and why it's needed for `1.6`?

Also could you make these cherry pick PRs to the `release/1.6`?

**Comment by AshkanAliabadi** [comment]
> Also could you make these cherry pick PRs to the `release/1.6`?

Right, they are optimizations for PyTorch Mobile and bug fixes for those optimizations.

**Comment by malfet** [comment]
> Requesting #40587

@dreiss can you please cherry-pick PR to `release/1.6`?

**Comment by AshkanAliabadi** [comment]
@malfet @dreiss I created cherry-pick PRs against release/1.6 for my PRs and David's.  Here's the list:

1. #40951
2. #40953
3. #40954
4. #40955
5. #40956

If you have any concerns lets follow up on the PRs individually to avoid cluttering this thread.  Thanks!

**Comment by jamesr66a** [comment]
Please merge https://github.com/pytorch/pytorch/pull/41013

**Comment by Krovatkin** [comment]
@malfet @dreiss could you please cherry-pick the following fixes to jit's shape analysis to 1.6

https://github.com/pytorch/pytorch/pull/41014
https://github.com/pytorch/pytorch/pull/41015
https://github.com/pytorch/pytorch/pull/41016

thank you!

**Comment by hx89** [comment]
Could you please merge the following fixes: 

#40890
#41023

Thanks!

**Comment by raghuramank100** [comment]
Could you merge this change that only affects the doc strings?
#41032

**Comment by ailzhang** [comment]
https://github.com/pytorch/pytorch/pull/41033

**Comment by mcarilli** [comment]
optional but minor and nice to have:
https://github.com/pytorch/pytorch/pull/41036

**Comment by ZolotukhinM** [comment]
Please merge #41017 after @Krovatkin's #41014, #41015, and #41016. The failures in #41017 should be fixed by these three PRs.

**Comment by eellison** [comment]
@malfet could you please merge the following shape analysis fixes ? thank you! 

https://github.com/pytorch/pytorch/pull/41044

**Comment by mrshenli** [comment]
#41089, port for #40963

This is a doc-only change, no impact on binary. cc @jlin27 @malfet

**Comment by malfet** [comment]
#41091 to fix PY37 check
#41090 to stop embeddeding protobuf symbols in every shared library

**Comment by mruberry** [comment]
https://github.com/pytorch/pytorch/pull/41098 fixes high priority #41047

**Comment by houseroad** [comment]
https://github.com/pytorch/pytorch/pull/41117 update the recompute_scale_factor to False, which keeps the consistency of scale_factor in C++ and Python. We started printing warning in 1.5, and told users we will turn this to False in 1.6. Also verified no impact on torchvision and detectron2 models with @fmassa and @ppwwyyxx

**Comment by rohan-varma** [comment]
https://github.com/pytorch/pytorch/pull/41151, Port of #40115, which fixes bug reported in https://github.com/pytorch/pytorch/issues/37790

**Comment by mrshenli** [comment]
#41156, port for #41108 

This is a doc-only change, no impact on binary. cc @jlin27 @malfet

**Comment by gchanan** [comment]
Placeholder for https://github.com/pytorch/pytorch/pull/41139/.

**Comment by gchanan** [comment]
Placeholder for https://github.com/pytorch/pytorch/pull/41155.

**Comment by gchanan** [comment]
Placeholder for https://github.com/pytorch/pytorch/pull/41152.

**Comment by anjali411** [comment]
Placeholder for https://github.com/pytorch/pytorch/pull/40927 which is a TH to ATen port for addmm. This also enables `matmul` for complex tensors which is important for the release.

**Comment by SsnL** [comment]
https://github.com/pytorch/pytorch/pull/41183

**Comment by SsnL** [comment]
https://github.com/pytorch/pytorch/pull/41184, port for https://github.com/pytorch/pytorch/pull/40344

**Comment by AshkanAliabadi** [comment]
https://github.com/pytorch/pytorch/pull/41190

@dreiss

**Comment by malfet** [comment]
> #41190
> 
> @dreiss

@AshkanAliabadi can you please explain in a few words, why update to pthreadpool is needed in 1.6?

**Comment by AshkanAliabadi** [comment]
> > #41190
> > @dreiss
> 
> @AshkanAliabadi can you please explain in a few words, why update to pthreadpool is needed in 1.6?

Sure.  A critical part of running mobile models efficiently is running ```torch.utils.mobile_optimizer.optimize_for_mobile``` on them on the host to perform a number of optimizations.  This workflow is not supported on Windows.  That is because XNNPACK only recently added Windows support, so updating XNNPACK is a perquisite of enabling that workflow.  XNNPACK itself though, is dependent on a number of other libraries, pthreadpool being one of them.  pthreadpool itself also enabled Windows support recently so updating pthreadpool too is a perquisite of that as well.  The rest of the dependencies (cpuinfo, fp16, fxdiv, and psimd) I have referenced previously above which you have already merged into this release branch.

More important than the submodule update in #41190 is the inclusion of a fix to a serious bug that had gone undetected through CI which was only brought to light after this library was updated.

All of the PRs in this series I have linked are part of an "all or none" package unfortunately, so we either want to have them all in, or revert them all.  If we can't get the XNNPACK update in to enable the said workflow, the series of commits you merged in this line (by me and David) bring noticeable performance improvements to mobile, but again, anything up to and including this PR is all or none.

**Comment by mcarilli** [comment]
small docstring clarification for `torch.cuda.amp`
https://github.com/pytorch/pytorch/pull/41204

**Comment by zou3519** [comment]
#41185, port for #41175, which is a documentation update / clarification

**Comment by jlin27** [comment]
modify sphinx writer to restore old documentation references
https://github.com/pytorch/pytorch/pull/39086

**Comment by malfet** [comment]
https://github.com/pytorch/pytorch/pull/41212 that prints a warning if pytorch is incompatible with current GPU

**Comment by anjali411** [comment]
Placeholder for Complex Autograd documentation: https://github.com/pytorch/pytorch/pull/41012

**Comment by lw** [comment]
Please merge #41229 (port of #41200), which is docs-only. Thanks!

**Comment by mcarilli** [comment]
docstring clarification for `torch.cuda.amp`
https://github.com/pytorch/pytorch/pull/41204
(@malfet Sibling PR against master https://github.com/pytorch/pytorch/pull/41203 is now merged so that's no longer a blocker)

**Comment by anjali411** [comment]
doc update/clarification https://github.com/pytorch/pytorch/pull/41263

**Comment by bzinodev** [comment]
please merge [#41271] (https://github.com/pytorch/pytorch/pull/41271)
This is a bug fix for torch.jit.freezing API

Sibling PR is landed in master[ #40718](https://github.com/pytorch/pytorch/pull/40718)

**Comment by ngimel** [comment]
See discussion in https://github.com/pytorch/pytorch/pull/40452, the user wants to see at least change to CUDAStream.h so that they can build 1.6 with cuda11. Let me know if you want a PR with change to CUDAStream.h change only, or if we should cherry-pick all of #40452

**Comment by Flamefire** [comment]
Requesting #41445 to be included which fixes a faulty test failing when running all tests. The fix is trivial

**Comment by anjali411** [comment]
Please merge doc note for complex: https://github.com/pytorch/pytorch/pull/41252

**Comment by jlin27** [comment]
Please merge for optimizer_for_mobile doc into python api root doc: https://github.com/pytorch/pytorch/pull/41491

**Comment by anjali411** [comment]
re ping to merge these doc PRs:
https://github.com/pytorch/pytorch/issues/40472#issuecomment-656424798
https://github.com/pytorch/pytorch/issues/40472#issuecomment-656857886
https://github.com/pytorch/pytorch/issues/40472#issuecomment-658923662

**Comment by ezyang** [comment]
Doc PR (edits header file, but only non-substantive changes; doesn't have to go in release build, just the branch before we build final cpp docs): https://github.com/pytorch/pytorch/pull/41602

**Comment by Flamefire** [comment]
Requesting to include:

- #41982 Fix to allow using system pybind11
- #41988 Fix flaky test
- #41989 Fix superflous download of psimd during configure
- #41991 Fix CMake warning which might become an error
- #41993 Fix failing test_tensorexpr
- #41994 Fix flaky rpc test due to missing shutdown
- #41995 Fix test failing when using at least 4 but less than 6 GPUs, see #41365 which is the open issue
- #41996 Fix test running wrongly on CPU instead of GPU failing when not built with MKL

All of those PRs have an already merged PR against master (link to that is in each of those PRs), fix an issue and are very small (sometimes single line change, pretty much trivial in all cases)

CC @lw @mruberry as the authors of most of the original PRs

**Comment by cyyever** [comment]
I need #41913

**Comment by Flamefire** [comment]
Requesting to merge #42010 which disables a test for the 1.6 branch which makes CI fail all PRs against the release/1.6 branch.   
Commit backported from master

**Comment by shihaoyin** [comment]
Please merge #41034

**Comment by seemethere** [comment]
Hey everyone! I've locked comments on this issue, we have a final RC for 1.6.0 and will therefore be no longer accepting cherry picks for 1.6.0

Thanks everyone for their submissions!

**Comment by seemethere** [comment]
`v1.6.0` has gone GA! Closing this issue, until the next release everyone!
